Introduced Spring boot actuators

LucaGiamminonni opened the pull request:
## References
Fixes #2877 and introduce the spring boot actuators in dspace. In particular, the endpoints that are enabled are /info and /health. 

Related to REST contract: https://github.com/DSpace/RestContract/pull/182
Related to Angular PR: https://github.com/DSpace/dspace-angular/pull/1637

## Description
It is possible to configure which endpoints to expose through the properties management.endpoints.web.exposure.include and management.endpoints.web.exposure.exclude (for more details https://docs.spring.io/spring-boot/docs/current/reference/html/actuator.html#actuator.endpoints.exposing). The status shown by the actuator health is calculated using the following indicators:

- datasource indicator
- solr indicator
- custom geoIp indicator that check the GeoIp configuration (can be disabled by setting the property management.health.geoIp.enabled = false)

The actuator root (by default /actuator) is displayed in the hal browser entry point. You can change the base path by configuring the management.endpoints.web.base-path property. 

The health endpoint is configured via the management.endpoint.health.show-details and management.endpoint.health.roles properties to show status details only to users who have an admin token.

## Instructions for Reviewers
To check the changes in this PR you can:
- open the hal browser entry point and see that the entry for the actuators is present
- by clicking on the actuator link you are redirected to the actuator root where the health and info endpoints are visible
- contacting the health endpoint not as an admin user only shows the status of the application, with no other details
- contacting the health endpoint as the admin user shows the status of the application with the details of the individual indicators

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [X] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

Commit:
[CST-5288] Introduced spring boot actuators

Commit:
[CST-5288] Added custom health indicator to check GeoIP configuration status

Commit:
[CST-5288] Made the actuator security configuration dynamic

LucaGiamminonni commented:
The application called **Spring Boot Admin Server** ( https://github.com/codecentric/spring-boot-admin ) can be used to test and view the information exposed by the actuators. To configure it, you can use the spring initializer page ( https://start.spring.io/ ) by adding **Codecentric's Spring Boot Admin (Server)** to the dependencies. On this project we must then add the @EnableAdminServer annotation to the main class.

You also need to configure the dspace application to be monitored by adding the spring-boot-admin-starter-client maven dependency; this would allow dspace to register itself on the server. Then you need to add the following properties on dspace:
spring.boot.admin.client.url -> the spring boot admin server url
spring.boot.admin.client.instance.serviceBaseUrl -> the base url of dspace itself

As a test I also recommend configuring dspace to make all the actuator information public:
management.endpoints.web.exposure.include = *
management.endpoint.health.show-details = always

By starting both applications you should be able to see dspace between applications registered on the server:
![image-20220211-092434](https://user-images.githubusercontent.com/16231083/154957261-fc7ea695-034b-4401-a2f4-a5e2680cd716.png)

Also it should be possible to see the details:
![image-20220211-092413](https://user-images.githubusercontent.com/16231083/154957275-e74c344d-067f-436b-9d29-e5b022998574.png)

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/HealthIndicatorsIT.java:
```diff
@@ -0,0 +1,72 @@
+
+        getClient(token).perform(get(HEALTH_PATH))
+            .andExpect(status().isServiceUnavailable())
+            .andExpect(jsonPath("$.status", is(Status.OUT_OF_SERVICE.getCode())))
```
It seems a bit odd to have the entire `/health` actuator display "OUT OF SERVICE" if only the `geoIp` check fails.  I'd consider `geoIp` to be optional, as not every site really uses Solr Statistics...and the site will function perfectly well even if `geoIP` is missing.

So, in this test, I can understand having the `geoIp` component respond with "OUT OF SERVICE".  But it seems odd for the entire status to then become "OUT OF SERVICE".

tdonohue reviewed (commented):
@LucaGiamminonni : Gave this a test and review today.  Overall, as we've discussed in meetings, I like the idea.  I feel this PR seems a bit "incomplete" currently, but it's a good start.

A few things that look to be missing to me:
1. I think the `info` actuator should return more info about the DSpace site (though, at least some of this information may need to be restricted to Administrators). For example, the following "information" was available in the old DSpace XMLUI Control Panel
    * Version of DSpace.
    * Possibly also the version of Java (as that was in the old Control Panel as well)
    * Values of key configurations (e.g. `dspace.dir`, `dspace.ui.url`, `dspace.server.url`, `solr.server`, `db.url`, `mail.server`, possibly also various configured email addresses, e.g. mail.admin. I'd also recommend including `rest.cors.allowed-origins` here.. and any other configs that we feel are useful to display in a control panel.)
    * (Optional, could be added later) We could also consider checking if OAI, SWORD(v1 and v2) and RDF endpoints are enabled. These are all just configs as well, e.g. `oai.enabled`, `sword-server.enabled`, etc.
2. The `health` actuator looks like a good start.  However, we may want to find a way to make some checks *optional* (e.g. see my inline comment about `geoIp`, which I think should be optional...if it fails it shouldn't cause the overall status to also be a failure).  We also might want to enhance the checks to include:
     * Checking all the Solr cores, not just the search core.

Overall, this is good work. Just wanted to note that we likely need to minimally make a few small enhancements here to better align with what we'd expect out of a Control Panel.

LucaGiamminonni commented:
Hi @tdonohue, thanks for your review. 
Regarding the info endpoint I can do the changes to add the information you suggested.
Regarding the health indicator and the geoIp, I would prefer that the geoIp status affects the global one because the health indicator in my opinion should be a tool that quickly gives you feedback on the status of the application: exclude from this status one of the health indicators could lead to not noticing any problems with the geoIp configuration (especially if we will have more indicators in the future). To mitigate the problem you describe I might suggest:
- disable the geoIp health indicator with management.health.geoIp.enabled = false
- create a new group of health indicator in which to include only the main indicators (https://docs.spring.io/spring-boot/docs/current/reference/html/actuator.html#actuator.endpoints.health.groups). In this way, the geoIp indicator could be excluded in order not to affect the general status.

tdonohue commented:
@LucaGiamminonni : Thanks for your response. I agree with your overall assessment that, when the `geoIp` health indicator is enabled, it should trigger the overall status to _change_. I just don't like that it currently makes the status become `OUT_OF_SERVICE` (as that status implies the site is down or not working at all...but we know a site can continue to function even if `geoIp` is itself `OUT_OF_SERVICE`).

So, I think your idea to create a new status between `UP` and `OUT_OF_SERVICE` is a good one, and it would address my feedback here.  It could be an `ISSUES_EXIST` status, or something similar, which implies the site may have issues that need addressing.  I also like that it's possible to simply *disable* the `geoIp` health indicator, if that issue is not a concern to the site...that'd obviously then ensure the `health` status would move to `UP`, even if `geoIp` is not configured or working. 

Overall, I think this all sounds good, and I like the direction to add in a new status.  I'm not sure `ISSUES_EXIST` is the proper name, so feel free to suggest a status you feel may be more appropriate.

Commit:
Merge remote-tracking branch '4science-bitbucket/main' into CST-5288

Commit:
[CST-5288] Improved health actuator

Commit:
[CST-5288] Improved info actuator

LucaGiamminonni commented:
Hi @tdonohue, thanks for your feedback. I updated the PR with the following changes:

- I introduced a new status UP_WITH_ISSUES which is placed between the status UP and OUT_OF_SERVICE; in this way, if the geoIp configuration has problems, the system is not shown as OUT_OF_SERVICE but with the status UP_WITH_ISSUES
- I added a health indicator for each core solr
- I added the information you suggested to the info endpoint and it should now return something like this:
`{
  "app": {
    "name": "DSpace at My University",
    "version": "7.3",
    "dir": "/d/install/dspace7",
    "url": "http://localhost:8080/server",
    "solr": {
      "server": "http://localhost:8983/solr",
      "prefix": ""
    },
    "db": "jdbc:postgresql://localhost:5432/dspace7",
    "mail": {
      "server": "smtp.example.com",
      "from-address": "dspace-noreply@myu.edu",
      "feedback-recipient": "dspace-help@myu.edu",
      "mail-admin": "dspace-help@myu.edu",
      "mail-helpdesk": "dspace-help@myu.edu",
      "alert-recipient": "dspace-help@myu.edu"
    },
    "cors": {
      "allowed-origins": "http://localhost:4000"
    },
    "ui": {
      "url": "http://localhost:4000"
    }
  }
}`

I haven't added the Java version information yet because there is a specific contributor for this purpose (https://github.com/spring-projects/spring-boot/blob/v2.6.5/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/info/JavaInfoContributor.java) which is only available since version 2.6 of Spring Boot. I saw that a spring boot upgrade is planned and so I was thinking of introducing it after that upgrade.

Commit:
[CST-5288] Fixed GeoIpHealthIndicatorTest

Commit:
[CST-5288] Disabled solr oai core health indicator during tests

LucaGiamminonni commented:
I had to disable aoi solr core health indicator on the test environment because I don't think there is a mock server for that core and so the build was working locally but not here on github.

Commit:
[CST-5288] Add CorsMapping for actuator endpoint

Commit:
[CST-5288] Fixed CORS configuration for actuator endpoints

LucaGiamminonni commented:
The CORS configuration for the actuator requires to configure a specific set of properties so I can't use a configuration similar to the one used for api and iiif endpoints. I had tried to configure cors for the actuators in Application.webMvcConfigurer().new WebMvcConfigurer() {...}.addCorsMappings(CorsRegistry) but that configuration was ignored. I had to add the properties management.endpoints.web.cors.* in the dspace.cfg to solve the problem

tdonohue commented on dspace/config/dspace.cfg:
```diff
@@ -1572,6 +1572,23 @@ request.item.type = all
+management.health.ping.enabled = false
+management.health.diskSpace.enabled = false
+
+management.endpoints.web.cors.allowed-origins = ${rest.cors.allowed-origins}
```
Again, this section needs comments to describe that it controls the CORS settings of ALL `/actuator/*` endpoints.  We also should make a note here that you'd want to sync these with the CORS settings in `org.dspace.app.rest.Application` (Sidenote: eventually we should probably just make them all one configuration to avoid duplicating the same thing in both places. But, that can be done in a future PR)

LucaGiamminonni replied:
Ok I will add a comment

tdonohue commented on dspace/config/modules/info-actuator.cfg:
```diff
@@ -0,0 +1,19 @@
+
+info.app.name = ${dspace.name}
+info.app.version = 7.3
```
We should avoid hardcoding a version here in configuration as this is not accurate.  We already report a programmatically determined version in `dspaceVersion` property from `/server/api`.  So, if this setting cannot be applied programmatically, then we should just remove `info.app.version` altogether...as it'll likely be wrong.

tdonohue commented on dspace/config/modules/info-actuator.cfg:
```diff
@@ -0,0 +1,19 @@
+info.app.mail.feedback-recipient = ${feedback.recipient}
+info.app.mail.mail-admin = ${mail.admin}
+info.app.mail.mail-helpdesk = ${mail.helpdesk}
+info.app.mail.alert-recipient = ${alert.recipient}
```
We should add some basic inline comments & docs here to let people know that you can remove one of these settings from being reported from the `/actuator/info` endpoint by just commenting it out & that you can add new settings by just creating a new key starting with "info.app".

tdonohue commented on dspace-api/src/test/data/dspaceFolder/config/local.cfg:
```diff
@@ -144,3 +144,5 @@ authentication-ip.Student = 6.6.6.6
 proxies.trusted.ipranges = 7.7.7.7
 proxies.trusted.include_ui_ip = true
+
+management.health.solrOai.enabled = false
```
Could we add a comment as to why this is needed?  It looked odd to me at first, but it looks like you are disabling the OAI health check because it will always fail in the test environment?

It's just good to add notes to other developers as to why some setting needs to be made in the "global" local.cfg for all tests.

LucaGiamminonni replied:
I had to disable this health indicator because there isn't a mock server for it and so the build was working locally but not here on github.

tdonohue reviewed (changes requested):
@LucaGiamminonni : Gave this another look today.  Looks great overall & it works as expected. I do have some minor suggestions inline though.  Mostly, I'd rather we create a single `actuator.cfg` config for all the actuator configurations.... I suspect people may want to edit them all at once instead & it'd just make sense to keep them together instead of separated.

A few other notes inline below...but I think it's all minor.  Beyond that, this looks good to me!

LucaGiamminonni commented on dspace/config/dspace.cfg:
```diff
@@ -1572,6 +1572,23 @@ request.item.type = all
 
+#---------------------------------------------------------------#
+#------------SPRING BOOT ACTUATOR CONFIGURATION-----------------#
+#---------------------------------------------------------------#
```
Yes, you are right, it is better to move these properties to a specific file. Thanks

Commit:
[CST-5288] Moved all spring boot actuator configurations in actuator.cfg

Commit:
[CST-5288] Added Order to the VersionInfoContributor

tdonohue commented:
@LucaGiamminonni : It looks like there are test failures now in `org.dspace.app.rest.example.ExampleControllerIT`.  I suspect it's something small (maybe a missing dependency or configuration).  Hopefully it's easy enough to clean up.  In any case, I can still give this an updated review/test in the meantime.

LucaGiamminonni commented:
Hi @tdonohue, the error seems to come from the fact that when the health indicator for authority's solr core is created the AuthoritySolrServiceImpl.getSolr () method tries to contact solr at localhost:8983. I don't know why this error occurs only now since it wasn't there in the previous commits. Any idea? is it correct that test causes an attempt to query solr at that address?

KevinVdV reviewed (changes requested):
I made an initial review of the codebase here, going to split it up into sections. But my overall impression is good.

### Info endpoint

The info endpoint which exposes certain configuration properties like (db url, solr server url, a lot of mails, ...) is fully public. I would advocate to place this behind an admin credential, we don't want to expose these things to bots.

### The spring boot admin webapp

I'm able to get the spring boot admin webapp to startup just fine BUT you mention that DSpace also needs an "spring-boot-admin-starter-client" dependency, which version did you try locally ? Because if I use the last one I run into issues with the "maven-enforcer-plugin" on spring versions.

### About the failing tests

I think I also know WHY the tests are failing, now that the "solr.authority.server" is enabled by default we should also properly mock the solr in the tests. Take a look at the MockAuthoritySolrServiceImpl class, this will set the solr to "Null".

We should mock this similar to the MockSolrSearchCore class.

atarix83 committed:
Merge branch 'main' into CST-5288

atarix83 committed:
[CST-5288] Fix issue after alignment with main branch

Commit:
[CST-5288] Fix actuators issues

Commit:
[CST-5288] Added admin security check on info endpoint

tdonohue commented on dspace/config/modules/actuator.cfg:
```diff
@@ -0,0 +1,54 @@
+
+info.app.ui.url = ${dspace.ui.url}
+
+actuator.info.version-attribute = version
```
Would it be possible to move the DSpace version under `info.app` in some manner?  In my [review of the frontend PR](https://github.com/DSpace/dspace-angular/pull/1637#pullrequestreview-963846321), I noted that it's currently out-of-place, and it has no context.  Technically, the DSpace version is a property of the "app".  But, currently it appears to be returned separately.

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/HealthIndicatorsIT.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public class HealthIndicatorsIT extends AbstractControllerIntegrationTest {
+
+    private static final String HEALTH_PATH = "/actuator/health";
```
While it is great that we have basic tests for the `/actuator/health` endpoint, it appears we are missing tests for `/actuator/info`.  We should be verifying that `/actuator/info` is only accessible to Admins, and that it returns basic settings that are configured by default in `actuator.cfg`.

tdonohue reviewed (changes requested):
@LucaGiamminonni : Overall this is looking great!  I'm nearly a +1, but I noticed a few minor things during my code review & my testing of this with the frontend PR.  See inline comments below.

Commit:
Merge remote-tracking branch '4science-bitbucket/main' into CST-5288

Commit:
[CST-5288] Info endpoint improvements

Commit:
Merge remote-tracking branch '4science-bitbucket/main' into CST-5288

Commit:
[CST-5288] Removed version check on InfoEndpointIT

KevinVdV reviewed (commented):
The code looks good to me but I would still like to do a full test with the boot webapp (I would like to verify that the authentication works & that the endpoints properly work in that app). I had some problems with the installation (see my previous comment [here](https://github.com/DSpace/DSpace/pull/8166#pullrequestreview-953577702)).

Could you provide a step by step guide in how to set that up ?

KevinVdV reviewed (changes requested):
I did an additional review using the angular UI. When everything works & REST isn't throwing any issues it works well.

But when I shutdown the solr instance the REST returns a 503 & so the ANGULAR UI doesn't work anymore. This was already found during the angular testing: https://github.com/DSpace/dspace-angular/pull/1637#issuecomment-1127954777

This should be fixed before we merge this into main.

FYI: This issue didn't occur when I removed my GeoLocation file, that shows down when it is gone.

Commit:
[CST-5288] Set 200 as http status for down and out_of_service health status

tdonohue commented on dspace/config/modules/actuator.cfg:
```diff
@@ -0,0 +1,57 @@
+
+## Configuration to set 200 as status of health http response when it is DOWN or OUT_OF_SERVICE
+management.endpoint.health.status.http-mapping.down = 200
+management.endpoint.health.status.http-mapping.out-of-service = 200
```
I'd recommend we add a note here to say:
"The DSpace UI requires these be set to 200 in order to support health status reports when services are down."

tdonohue reviewed (approved):
@LucaGiamminonni : 👍 Thanks for the updates! This looks good to me.  I've also tested it today with the UI and verified that the UI is now able to provide a detailed health status when services are down. For example, here's what you see when Solr is down:

![Solrdown](https://user-images.githubusercontent.com/483997/168874229-e2dd0fc9-9f06-45f8-abf1-e201227a1c28.png)

Overall, this looks great.  That said, I have **one minor feedback inline below**.. I'd appreciate it if we can make this small update before it's merged.

Commit:
[CST-5288] Improved health status comment

KevinVdV reviewed (approved):
The change with the status code works well. I was able to test that the solr can be shut down while the DSpace is running without killing the entire health endpoint

Approving the rest side of things.

Thanks for the hard work @LucaGiamminonni !
